### PR TITLE
Oneshot mapper

### PR DESCRIPTION
**Pull request by rdaly525**
depends on #108 

This is the full code for doing the general oneshot automatic mapping. It is somewhat based on code from the 'paper' branch, but fixes the family issue and is updated to do the single call to the SMT solver.

I would advise reading through the peak/mapper/README first, and going through the PR by commits as there is a lot of code.

In summary, the mapper constructs a single qualified Sum of Products expression which it seems the SMT solver can handle quite easily for this particular example.

### CODE REVIEW

File: peak/mapper/README.txt [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+The term 'path' refers to a tuple describing the 'path' to a particular node in the adt tree in terms of product/tuple/sum fields. For example:
+
+class A(Product):
+    a=T0
+    b=T1
+class B(Product):
+    a=Sum[A,T2]
+    b=T3
+
+Here are the complete list of paths for B (and their corresponding types)
```
Comment (cdonovick) [inline peak/mapper/README.txt]: Strings are getattr accessed while types are getitem accessed?

File: peak/mapper/README.txt [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+Form 1:
+  sum choices: [('a',) -> A]
+  leaf nodes:
+     ('a', A, 'a') -> T0
+     ('a', A, 'b') -> T1
+     ('b',) -> Bit
+
+Form 2:
+  sum choices: [('a',) -> T]
```
Comment (cdonovick) [inline peak/mapper/README.txt]: T2

File: peak/mapper/README.txt [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+family closure (often abbreviated as 'fc') is a function that takes in one argument (the family) and returns a Peak class defined in terms of that family. These functions should always be cached.
+
+---------------input_t/output_t--------------
+The automatic mapper requires any peak class to annotate the names and types of all inputs and outputs. This is accomplished using the 'name_outputs' decorator on the __call__ method. This gives everyone access to the adt type which represents the input (input_t) and the adt type represenitng the output (output_t).
+
+For example:
+
+class PE(Peak):
+   @name_outputs(o1=T3,o2=T4)
+   def __call__(self, a=T1,b=T2):
```
Comment (cdonovick) [inline peak/mapper/README.txt]: do you mean `a: T1, b: T2`?

Comment (rdaly525) [reply peak/mapper/README.txt]: Yeah, typo

**Comment by phanrahan** [comment]
What branches in other repos do I need to run the tests? None of them work for me.

**Comment by rdaly525** [comment]
now depends on https://github.com/leonardt/hwtypes/pull/106 

@phanrahan try ruunning this branch based off hwtypes:rebind_type

**Comment by phanrahan** [comment]
I was able to run the tests. 

I think we should do the merge as soon as possible. Just make sure the merge doesn't break any of the current tests, including in the repos that depend on peak. Most important is lassen and garnet.

**Comment by phanrahan** [comment]
Another suggestion based on my discussion with @rdaly525. Let's write a style guide which outlines the best practices for wiring peak programs.

File: tests/test_mapper_utils.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    ]
+    expected_paths = [
+        (0, A,'a'),
+        (0, A,'b'),
+        (0, SBV[8]),
+        (0, SBit),
+        (0, _TAG),
+        (0, Match),
+        (1,'a'),
+        (1,'b'),
```
Comment (cdonovick) [inline tests/test_mapper_utils.py]: white space

File: tests/test_mapper_utils.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+        #expected_paths should be exactly in varmap
+        assert len(expected_paths) == len(varmap)
+        for path in expected_paths:
+            assert path in varmap
+
+        #Matches should be consistent with forms
+        for form in forms:
+            assert form.path_dict in expected_forms
+            for path,field in form.path_dict.items():
```
Comment (cdonovick) [inline tests/test_mapper_utils.py]: whitespace

File: peak/mapper/mapper.py [hunk 6]
```diff
@@ -1,148 +1,254 @@
...
+        return IRMapper(self, ir_fc)
+
+class IRMapper(SMTMapper):
+    def __init__(self, archmapper, ir_fc):
+        super().__init__(ir_fc)
+        #For now assume that ir input forms and ir output forms is just 1
+        if self.num_input_forms > 1:
+            raise NotImplementedError("NYI, multiple ir input forms")
+        if self.num_output_forms > 1:
+            raise NotImplementedError("NYI, multiple ir output forms")
```
Comment (cdonovick) [inline peak/mapper/mapper.py]: rm NYI

File: peak/mapper/mapper.py [hunk 7]
```diff
@@ -1,148 +1,254 @@
...
+        self.num_output_forms = num_output_forms
+        self.num_input_forms = num_input_forms
+        self.input_varmap = input_varmap
+
+
+class ArchMapper(SMTMapper):
+    def __init__(self, arch_fc):
+        super().__init__(arch_fc)
+        if self.num_output_forms > 1:
+            raise NotImplementedError("NYI, multiple ir output forms")
```
Comment (cdonovick) [inline peak/mapper/mapper.py]: NYI is redundant

File: peak/mapper/mapper.py [hunk 8]
```diff
@@ -1,148 +1,254 @@
...
+
+    def process_ir_instruction(self, ir_fc):
+        return IRMapper(self, ir_fc)
+
+class IRMapper(SMTMapper):
+    def __init__(self, archmapper, ir_fc):
+        super().__init__(ir_fc)
+        #For now assume that ir input forms and ir output forms is just 1
+        if self.num_input_forms > 1:
+            raise NotImplementedError("NYI, multiple ir input forms")
```
Comment (cdonovick) [inline peak/mapper/mapper.py]: rm NYI

File: peak/mapper/mapper.py [hunk 9]
```diff
@@ -1,148 +1,254 @@
...
+        form_var = archmapper.input_form_var
+
+        #Create the form_conditions (preconditions) based off of the arch_forms
+        #[input_form_idx]
+        form_conditions = []
+        for fi, form in enumerate(archmapper.input_forms):
+            #form_condition represents the & of all the appropriate matche
+            conditions = [form_var == 2**fi]
+            for path, choice in form.path_dict.items():
+                match_path = path + (Match, )
```
Comment (cdonovick) [inline peak/mapper/mapper.py]: whitespace

File: peak/mapper/utils.py [hunk 10]
```diff
@@ -0,0 +1,339 @@
...
+        assembler = assembler_t(adt_t)
+        #Create _TAG
+        if value is None:
+            tag = SMTBitVector[assembler.tag_width]()
+        else:
+            tag = value[_TAG]
+
+        forms = []
+        varmap = {}
+        varmap[path + (_TAG, )] = tag
```
Comment (cdonovick) [inline peak/mapper/utils.py]: ws

File: peak/mapper/utils.py [hunk 11]
```diff
@@ -0,0 +1,339 @@
...
+        varmap[path + (Match, )] = {}
```
Comment (cdonovick) [inline peak/mapper/utils.py]: ws

File: peak/mapper/utils.py [hunk 12]
```diff
@@ -0,0 +1,339 @@
...
+        for field in adt_t.fields:
+            #field_tag_value = assembler.assemble_tag(field, bv_t)
+            #tag_match = (tag==field_tag_value)
+            sub_aadt_t = aadt_t[field]
+            if value is None:
+                sub_value = None
+            else:
+                sub_value = value[field].value
+            sub_forms, sub_varmap = self(sub_aadt_t, path=path + (field, ), value=sub_value)
```
Comment (cdonovick) [inline peak/mapper/utils.py]: ws after tailing comma

File: peak/mapper/utils.py [hunk 13]
```diff
@@ -0,0 +1,339 @@
...
+            #update sub_forms with current match path
+            for sub_form in sub_forms:
+                assert path not in sub_form.path_dict
+                path_dict = {path:field, **sub_form.path_dict}
+                if value is None:
+                    form_value = aadt_t.from_fields(field, sub_form.value, tag_bv=tag)
+                else:
+                    form_value = value
+                forms.append(Form(value=form_value, path_dict=path_dict, varmap=sub_form.varmap))
+            varmap[path + (Match, )][field] = forms[0].value[field].match
```
Comment (cdonovick) [inline peak/mapper/utils.py]: ws

File: peak/mapper/utils.py [hunk 14]
```diff
@@ -0,0 +1,339 @@
...
+        forms_to_product = []
+        #Needed to guarentee order is consistent
+        adt_items =  list(adt_t.field_dict.items())
+        for field_name, field in adt_items:
+            sub_aadt_t = aadt_t[field_name]
+            if value is None:
+                sub_value = None
+            else:
+                sub_value = value[field_name]
+            sub_forms, sub_varmap = self(sub_aadt_t, path=path + (field_name, ), value=sub_value)
```
Comment (cdonovick) [inline peak/mapper/utils.py]: ws

File: peak/mapper/utils.py [hunk 15]
```diff
@@ -0,0 +1,339 @@
...
+
+        forms_to_product = []
+        for _idx, (idx, field) in enumerate(adt_t.field_dict.items()):
+            assert _idx == idx
+            sub_aadt_t = aadt_t[idx]
+            if value is None:
+                sub_value = None
+            else:
+                sub_value = value[idx]
+            sub_forms, sub_varmap = self(sub_aadt_t, path=path + (idx, ), value=sub_value)
```
Comment (cdonovick) [inline peak/mapper/utils.py]: ws

File: peak/mapper/utils.py [hunk 16]
```diff
@@ -0,0 +1,339 @@
...
+        simplified_binding = self(sub_aadt_t, sub_binding)
+
+        #Completely simplified if only one binding and ir_path is a value instead of a path
+        if len(simplified_binding)==1 and not isinstance(simplified_binding[0][0], tuple):
+            sub_value, arch_path = simplified_binding[0]
+            assert arch_path is ()
+            value = aadt_t.from_fields(sub_field, sub_value)
+            return [(value, ())]
+        else:
+            return [(ir_path, (sub_field, ) + arch_path) for ir_path, arch_path in simplified_binding]
```
Comment (cdonovick) [inline peak/mapper/utils.py]: ws

File: peak/mapper/utils.py [hunk 17]
```diff
@@ -0,0 +1,339 @@
...
+            if is_product:
+                value = aadt_t.from_fields(**value_dict)
+            else: #is tuple
+                values = [v for _, v in sorted(value_dict.items())]
+                value = aadt_t.from_fields(*values)
+            return [(value, ())]
+        else:
+            ret_binding = []
+            for sub_field, binding in simplified_binding_dict.items():
+                ret_binding += [(ir_path, (sub_field, ) + arch_path) for ir_path, arch_path in binding]
```
Comment (cdonovick) [inline peak/mapper/utils.py]: ws

**Comment by rdaly525** [comment]
@cdonovick, any other changes requested? Id like to merge this when GarnetFlow is completed
